core/trino-main/src/main/java/io/trino/operator/DeduplicatingDirectExchangeBuffer.java

```diff
@@ -385,29 +405,428 @@ public synchronized void close()
+        private final QueryId queryId;
+        private final ExchangeId exchangeId;
+
+        private final long pageBufferCapacityInBytes;
+        private final ListMultimap<TaskId, Slice> pageBuffer = ArrayListMultimap.create();
+        private long pageBufferRetainedSizeInBytes;
+
+        private ExchangeManager exchangeManager;
+        private Exchange exchange;
+        private ExchangeSinkInstanceHandle sinkInstanceHandle;
+        private ExchangeSink exchangeSink;
+        private SliceOutput writeBuffer;
+
+        private int bufferedPageCount;
+        private long spilledBytes;
+        private int spilledPageCount;
+
+        private boolean inputFinished;
+        private boolean closed;
+
+        private PageBuffer(
+                ExchangeManagerRegistry exchangeManagerRegistry,
+                QueryId queryId,
+                ExchangeId exchangeId,
+                DataSize pageBufferCapacity)
+        {
+            this.exchangeManagerRegistry = requireNonNull(exchangeManagerRegistry, "exchangeManagerRegistry is null");
+            this.queryId = requireNonNull(queryId, "queryId is null");
+            this.exchangeId = requireNonNull(exchangeId, "exchangeId is null");
+            this.pageBufferCapacityInBytes = requireNonNull(pageBufferCapacity, "pageBufferCapacity is null").toBytes();
+        }
+
+        public void addPages(TaskId taskId, List<Slice> pages)
+        {
+            if (closed) {
+                return;
+            }
+
+            if (inputFinished) {
+                // ignore extra pages after input is marked as finished
+                return;
+            }
+
+            long pagesRetainedSizeInBytes = getRetainedSizeInBytes(pages);
+            if (exchangeSink == null && pageBufferRetainedSizeInBytes + pagesRetainedSizeInBytes <= pageBufferCapacityInBytes) {
+                pageBuffer.putAll(taskId, pages);
+                pageBufferRetainedSizeInBytes += pagesRetainedSizeInBytes;
+                bufferedPageCount += pages.size();
+                return;
+            }
+
+            if (exchangeSink == null) {
+                verify(exchangeManager == null, "exchangeManager is not expected to be initialized");
+                verify(exchange == null, "exchange is not expected to be initialized");
+                verify(sinkInstanceHandle == null, "sinkInstanceHandle is not expected to be initialized");
+                verify(writeBuffer == null, "writeBuffer is not expected to be initialized");
+
+                try {
+                    exchangeManager = exchangeManagerRegistry.getExchangeManager();
+                }
+                catch (TrinoException e) {
+                    if (EXCHANGE_MANAGER_NOT_CONFIGURED.toErrorCode().equals(e.getErrorCode())) {
+                        throw new TrinoException(EXCHANGE_MANAGER_NOT_CONFIGURED, "Exchange manager must be configured for failure recovery capabilities to be fully functional");
```

Comment:
Why do you need to rethrow changing message. Is the original one poor - maybe we should just change it?

Also it would be nice to verify configuration earlier that when we really need to use the spooling exchange. Otherwise we may not notice configuration problem for very long time until we hit the affected codepath.
Actually, I believe it is being validated in some other place, but then it would be nice to add a comment that this is the case.

Reply:
The default message just says "Exchange manager is not configured". The idea was to provide a message that is more actionable and easier to understand. 

As an afterthought the alternative message is not very actionable either. I first was thinking about changing the message to:

> Capacity of the deduplication buffer required for failure recovery to function has been exceeded. Please configure exchange manager to allow the buffer to spill to an external storage.

But then I thought that the message of the error thrown by the `ExchangeManager` should be more actionable in general. 

I'm going to remove this overwrite of a message and improve the message in the `ExchangeManager` as a separate commit.

Reply:
>  Otherwise we may not notice configuration problem for very long time until we hit the affected codepath.

I'm not sure if we should always try to eagerly disallow failure recovery when the ExchangeManager is not configured. For example query level retries may function properly if the buffer capacity is never exceeded. For task level retries some queries (for example simple selects) may function properly as well.

Reply:
For task-level retries the set of working queries is so limited that I strongly feel we should check at the very beginning that exchange manager is configured.

For query-level retries ... maybe not. Yet the cost of configuring it so it uses local disk is small. And then it is easier for Trino users to know that when they configured a cluster and it works for some test queries, it will actually work for other ones too. 
On the other hand it makes life harder for people who run Trino with disk-less machines.